File: daemon/php/giana.php
<?PHP
// =============================================================================
// GIANA Framework | Home Automation Made Easy. (LAMP || WAMP) + Arduino UNO r3.
// =============================================================================
// Copyright (C) 2013 Federico Pfaffendorf (www.federicopfaffendorf.com.ar)
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// any later version. 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/gpl.txt
// =============================================================================
// Please, use it under PHP 5 or greater.
// =============================================================================
include_once("inc/php_serial.class.php");
// =============================================================================
// CLASS Context
// =============================================================================
class Context
{
	
	// ===========================================================================
	// Properties
	// ===========================================================================

	public static $gianas = array();

  // -------------------------------------------------------------------------
  // Change this to match your services root URI
  // -------------------------------------------------------------------------
	const servicesRootURI = "http://localhost/giana/srv/";				// |
	// -------------------------------------------------------------------------		
	
  // -------------------------------------------------------------------------
  // Change this random string, here and in your security.config.php
  // -------------------------------------------------------------------------
	const SALT = "0G9UFVc8VQTCs5qosA6g";																		// |
	// -------------------------------------------------------------------------		

  // -------------------------------------------------------------------------
  // Change this match your Arduino baud rate
  // -------------------------------------------------------------------------
	const BaudRate = 9600;																									// |
	// -------------------------------------------------------------------------		
	
	// ===========================================================================
	// Methods
	// ===========================================================================

	public static function initialize()
	{

		// -------------------------------------------------------------------------
    // Add as many Gianas as Arduinos you have !
    // -------------------------------------------------------------------------
		Context::$gianas[] = new Giana("/dev/ttyACM0");													// |
		// -------------------------------------------------------------------------		
		
	}
	
	public static function getServicesRootURI()
	{
	
		return Context::servicesRootURI .
			(substr (Context::servicesRootURI, 
							strlen(Context::servicesRootURI) - 1, 1) == "/" ? "" : "/");
							
	}
	
	public static function getGianaAt($port)
	{
		foreach(Context::$gianas as $giana)
			if ($giana->port == $port)
				return $giana;
		Log::this("There is no Giana at port " . $port . ".", false);
		return null;
	}
	
}
// =============================================================================
// MAIN
// =============================================================================
{

	// =============================================================================
	Context::initialize();
	// =============================================================================

	// =============================================================================
	Log::this("Program started.", false);
	// =============================================================================

	// =============================================================================
	// Schedule
	// =============================================================================
	$schedule = Schedule::get();
	if ($schedule != null)
	{
			$giana = Context::getGianaAt($schedule->port);
			if ($giana != null)
					$giana->performSchedule($schedule);
			else
					$schedule->done(true);
	}	
	// =============================================================================
	// Request
	// =============================================================================
	$request = Request::get();
	if ($request != null)
	{
		$giana = Context::getGianaAt($request->port);
		if ($giana != null)
				$giana->performRequest($request);
		else
		{
			$response = new Response(
					$request->id,
					Response::Error
				);
			$response->post();
		}
	}
	// =============================================================================

	// =============================================================================
	sleep(1);	
	// =============================================================================
	
	// =============================================================================
	Log::this("Program ended.", false);
	// =============================================================================

}
// =============================================================================
// CLASS Giana
// =============================================================================
class Giana 
{

	// ===========================================================================
	// Properties
	// ===========================================================================

	public $port;

	// ===========================================================================
	// Constructors
	// ===========================================================================
	
	function __construct($port)
	{
		$this->port = $port;
	}
	
	// ===========================================================================
	// Public Methods
	// ===========================================================================

	public function performSchedule($schedule)
	{
	
		$value = $this->scheduleConditions($schedule->conditions);
		
		$request = new Request (
			null, $schedule->port, Action::Set, $schedule->pinType,
			$schedule->pinNumber, 
			$value ? $schedule->trueValue : $schedule->falseValue
		);
		
		set($request);
		
		$schedule->done(false);
	
	}
	
	public function performRequest($request)
	{

		$response = null;
	
		if ($request->action == Action::Get)
			$response = $this->get($request);
		else
			$response = $this->set($request);

		if ($response != null)
			$response->post();
		else
		{
			$response = new Response(
					$request->id,
					Response::Error
				);
			$response->post();
		}	
			
	}
	
	// ===========================================================================
	// Private Methods
	// ===========================================================================
	
	private function scheduleConditions($conditions)
	{
	
		foreach ($conditions as $condition)
		{
		
			$request = new Request (
				null, $condition->port, Action::Get, $condition->pinType,
				$condition->pinNumber, 0				
			);
			
			$response = get($request);
			
			switch($condition->sign)
			{
			
				case '>':
						if ($response->value < $condition->rawValue)
							return false;
					break;
			
				case '<':
						if ($response->value > $condition->rawValue)
							return false;
					break;
			
				case '=':
						if ($response->value != $condition->rawValue)
							return false;
					break;
			
			}
		
		}
	
		return true;
	
	}
	
	private function get($request)
	{
	
		$message = "<G";
		$message .= $request->pinType;
		if ($request->pinNumber < 10) $message .= "0";
		$message .= $request->pinNumber;
		$message .= ">";

		$serial = new phpSerial();
		$serial->deviceSet($request->port); 
		$serial->confBaudRate(Context::BaudRate); 		
		$serial->deviceOpen(); 
		$serial->sendMessage($message); 

		sleep(1);

		$message = $serial->readPort(); 

		$serial->deviceClose(); 
		
		$response = new Response (
				$request->id,
				(int)substr($message, 4, ($request->pinType == PinType::Digital ? 1 : 4))
			);
	
		return $response;
	
	}

	private function set($request)
	{
	
		$message = "<SD";
		if ($request->pinNumber < 10) $message .= "0";
		$message .= $request->pinNumber;
		$message .= $request->value;
		$message .= ">";
		
		$serial = new phpSerial();
		$serial->deviceSet($request->port); 
		$serial->confBaudRate(Context::BaudRate); 		
		$serial->deviceOpen(); 
		
		$serial->sendMessage($message); 
		
		sleep(1);

		$message = $serial->readPort(); 

		$serial->deviceClose(); 
	
		$response = new Response (
				$request->id,
				substr($message, 6, 2)
			);
	
		return $response;
	
	}
	
}
// =============================================================================
// CLASS Schedule
// =============================================================================
class Schedule 
{

	// ===========================================================================
	// Properties
	// ===========================================================================

	public $id;
	public $port;
	public $pinType;
	public $pinNumber;
	public $trueValue;
	public $falseValue;
	public $conditions = array();

	// ===========================================================================
	// Methods
	// ===========================================================================

	public static function get()
	{

		$domDocument;
	
		try {
		
			$domDocument =		
				REST::this(
					Context::getServicesRootURI() . "schedule/", 
					"GET");
				
		} 
		catch (Exception $exception)
		{
		
			Log::this($exception->getMessage(), true);
			return null;
			
		}

		if ($domDocument->getElementsByTagName("device")->length > 0)
		{
		
			$schedule = new Schedule();
		
			$schedule->id = $domDocument->getElementsByTagName("schedule")->item(0)->getAttribute("id");
			$schedule->port = $domDocument->getElementsByTagName("port")->item(0)->nodeValue;
			$schedule->pinType = $domDocument->getElementsByTagName("pinType")->item(0)->nodeValue;
			$schedule->pinNumber = $domDocument->getElementsByTagName("pinNumber")->item(0)->nodeValue;
			$schedule->trueValue = $domDocument->getElementsByTagName("trueValue")->item(0)->nodeValue;
			$schedule->falseValue = $domDocument->getElementsByTagName("falseValue")->item(0)->nodeValue;
		
			return $schedule;
		
		}
		
		return null;
	
	}
	
	public function done($error)
	{
	
		try {
		
			REST::this(
				Context::getServicesRootURI() . "schedule/?scheduleId=" . $this->id .
				"&error=" . ($error ? "Y" : "N"), 
				"POST");
				
		} 
		catch (Exception $exception)
		{
		
			Log::this($exception->getMessage(), true);
			
		}
		
	}
	
}
// =============================================================================
// CLASS Condition
// =============================================================================
class Condition
{

	// ===========================================================================
	// Properties
	// ===========================================================================

	public $port;
	public $pinType;
	public $pinNumber;
	public $sign;
	public $rawValue;

	// ===========================================================================
	// Constructors
	// ===========================================================================

	function __construct ($port, $pinType, $pinNumber, $sign, $rawValue)
	{
	
		$this->port = $port;
		$this->pinType = $pinType;
		$this->pinNumber = $pinNumber;
		$this->sign = $sign;
		$this->rawValue = $rawValue;
	
	}
	
}
// =============================================================================
// CLASS Request
// =============================================================================
class Request
{

	// ===========================================================================
	// Properties
	// ===========================================================================
	
	public $id;
	public $port;
	public $action;
	public $pinType;
	public $pinNumber;
	public $value;

	// ===========================================================================
	// Constructors
	// ===========================================================================
	
	function __construct($id, $port, $action, $pinType, $pinNumber, $value)
	{
	
		$this->id = $id;
		$this->port = $port;
		$this->action = $action;
		$this->pinType = $pinType;
		$this->pinNumber = $pinNumber;
		$this->value = $value;
		
	}

	// ===========================================================================
	// Public Methods
	// ===========================================================================
	
	public static function get()
	{
	
		$domDocument;
	
		try {
		
			$domDocument =
				REST::this(
					Context::getServicesRootURI() . "request/", "GET");
				
		} 
		catch (Exception $exception)
		{
		
			Log::this($exception->getMessage(), true);
			return null;
			
		}
	
		if ($domDocument->getElementsByTagName("action")->length > 0)
		{
		
			$request = new Request(
				$domDocument->getElementsByTagName("requestId")->item(0)->nodeValue,
				$domDocument->getElementsByTagName("port")->item(0)->nodeValue,
				$domDocument->getElementsByTagName("action")->item(0)->nodeValue,
				$domDocument->getElementsByTagName("pinType")->item(0)->nodeValue,
				$domDocument->getElementsByTagName("pinNumber")->item(0)->nodeValue,
				$domDocument->getElementsByTagName("value")->item(0)->nodeValue
			);
		
			return $request;
		
		}
	
		return null;
	
	}
	
}
// =============================================================================
// CLASS Response
// =============================================================================
class Response
{
	
	// ===========================================================================
	// Properties
	// ===========================================================================
	
	public $requestId;
	public $value;
	
	const Error = "ER";
	const Ok = "OK";
	
	// ===========================================================================
	// Constructors
	// ===========================================================================

	function __construct($requestId, $value)
	{
		$this->requestId = $requestId;
		$this->value = $value;
	}

	// ===========================================================================
	// Public Methods
	// ===========================================================================

	public function post()
	{

		try {

			REST::this(
				Context::getServicesRootURI() . 
				"response/?requestId=" . $this->requestId .
				"&value=" . $this->value, "POST");
																	 
		} 
		catch (Exception $exception)
		{
		
			Log::this($exception->getMessage(), true);
		
		}
																	 
	}
	
}
// =============================================================================
// CLASS REST
// =============================================================================
class REST
{

	// =========================================================================
	// Public Methods
	// =========================================================================

	public static function this($uri, $method)
	{
	
		if (strpos($uri, "?") === false)
			$uri .= "?SALT=" . Context::SALT;
	  else
			$uri .= "&SALT=" . Context::SALT;

		$curl = curl_init();			
		
		switch ($method)
    {
      case "POST":
				curl_setopt($curl, CURLOPT_POST, 1);
				curl_setopt($curl, CURLOPT_POSTFIELDS, "");
				break;
      case "PUT":
          curl_setopt($curl, CURLOPT_PUT, 1);
				break;
    }
		
		curl_setopt($curl, CURLOPT_URL, $uri);
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);

		$domDocument = new DomDocument();
		//$respuesta=curl_exec($curl);
		//echo $respuesta;
		//$domDocument->loadXML($respuesta);
		$domDocument->loadXML(curl_exec($curl));
		
    return $domDocument;
		
	}

}
// =============================================================================
// CLASS Log
// =============================================================================
class Log
{

	// =========================================================================
	// Public Methods
	// =========================================================================

	public static function this($message, $exception)
	{
		echo(
			date("Y-m-d H:i:s") . " | " . ($exception ? "[EXCEPTION]" : "") . 
			$message . "\r\n"
		);
	}

}
// =============================================================================
// CLASS Action
// =============================================================================
class Action
{
	const Get = "G";
	const Set = "S";
}
// =============================================================================
// CLASS PinType
// =============================================================================
class PinType
{
	const Analogue = "A";
	const Digital = "D";
}
// =============================================================================
?>

File: daemon/NET/Giana.cs
// =============================================================================
// GIANA Framework | Home Automation Made Easy. (LAMP || WAMP) + Arduino UNO r3.
// =============================================================================
// Copyright (C) 2013 Federico Pfaffendorf (www.federicopfaffendorf.com.ar)
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// any later version. 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/gpl.txt
// =============================================================================
// Please, compile using .NET Framework 2.0 or greater.
// =============================================================================
using System;
using System.IO;
using System.Xml;
using System.Net;
using System.Text;
using System.IO.Ports;
using System.Threading;
using System.Collections.Generic;
using System.Collections.ObjectModel;
// =============================================================================
// CLASS Context
// =============================================================================
static class Context
{

    #region Properties

    // -------------------------------------------------------------------------
    // Add as many Gianas as Arduinos you have !
    // -------------------------------------------------------------------------
    private static Giana[] Gianas = new Giana[] {                           // |
                                        new Giana(                          // |
                                            new SerialPort("COM13", 9600)   // |
                                        )                                   // |
                                    };                                      // |
    // -------------------------------------------------------------------------

    // -------------------------------------------------------------------------
    // Change this to match your services root URI
    // -------------------------------------------------------------------------
    private static string _ServicesRootURI =                                // |
        "http://localhost:8888/giana/srv/";                                 // |
    // -------------------------------------------------------------------------

    public static string ServicesRootURI
    {
        get
        {
            return _ServicesRootURI + (
                _ServicesRootURI.EndsWith("/") ? "" : "/"
                );
        }
    }

    // -------------------------------------------------------------------------
    // Change this random string, here and in your security.config.php
    // -------------------------------------------------------------------------
    private static string _SALT =                                           // |
        "0G9UFVc8VQTCs5qosA6g";                                             // |
    // -------------------------------------------------------------------------

    public static string SALT
    {
        get
        {
            return _SALT;
        }
    }

    #endregion

    #region Methods

    public static Giana GetGianaAt(string port)
    {
        foreach (Giana giana in Gianas)
            if (giana.Port == port)
            {
                return giana;
            }
        Log.This("There is no Giana at port " + port + ".");
        return null;
    }

    #endregion

}
// =============================================================================
// CLASS Program
// =============================================================================
class Program
{

    #region Main

    public static void Main()
    {

        Log.This("Program started.");

        while (true)
        {

            Schedule schedule;
            Request request;
            Giana giana;

            #region Schedule

            schedule = Schedule.Get();

            if (schedule != null)
            {
                giana = Context.GetGianaAt(schedule.Port);
                if (giana != null)
                    giana.Perform(schedule);
                else
                    schedule.Done(true);
            }

            #endregion

            #region Request

            request = Request.Get();

            if (request != null)
            {
                giana = Context.GetGianaAt(request.Port);
                if (giana != null)
                    giana.Perform(request);
                else
                    new Response(
                            (int)request.Id,
                            Response.Result.Error
                        ).Post();
            }

            #endregion

            Thread.Sleep(1000);

        }

    }

    #endregion

}
// =============================================================================
// CLASS Giana
// =============================================================================
class Giana
{

    #region Properties

    private SerialPort serialPort = null;

    public string Port
    {
        get { return serialPort.PortName; }
    }

    #endregion

    #region Constructors

    public Giana(SerialPort serialPort)
    {

        this.serialPort = serialPort;
        if (!serialPort.IsOpen)
            try
            {
                serialPort.Open();
            }
            catch (Exception exception)
            {
                Log.This(exception);
                Environment.Exit(1);
            }

    }

    #endregion

    #region Public Methods

    public void Perform(Schedule schedule)
    {

        bool value = ScheduleConditions(schedule.Conditions);

        Request request = new Request(
            null,
            schedule.Port, Action.Set, schedule.PinType,
            schedule.PinNumber,
            (value ? schedule.TrueValue : schedule.FalseValue)
            );

        Set(request);

        schedule.Done(false);

    }

    public void Perform(Request request)
    {

        Response response;

        if (request.Action == Action.Get)
            response = this.Get(request);
        else
            response = this.Set(request);

        if (response != null)
            response.Post();
        else
        {
            new Response(
                    request.Id,
                    Response.Result.Error
                ).Post();
        }

    }

    #endregion

    #region Private Methods

    private bool ScheduleConditions(ReadOnlyCollection<Schedule.Condition> conditions)
    {

        foreach (Schedule.Condition condition in conditions)
        {

            Request request = new Request(
                    null,
                    condition.Port, Action.Get, condition.PinType,
                    condition.PinNumber, 0
                );

            Response response = Get(request);

            switch (condition.Sign)
            {

                case '>':
                    if (int.Parse(response.Value) < condition.RawValue)
                        return false;
                    break;

                case '<':
                    if (int.Parse(response.Value) > condition.RawValue)
                        return false;
                    break;

                case '=':
                    if (int.Parse(response.Value) != condition.RawValue)
                        return false;
                    break;

            }

        }

        return true;

    }

    private Response Get(Request request)
    {

        string message = "<G";
        if (request.PinType == PinType.Digital) message += "D";
        else message += "A";
        if (request.PinNumber < 10) message += "0";
        message += request.PinNumber.ToString();
        message += ">";

        try
        {
            serialPort.Write(message);
        }
        catch (Exception exception)
        {
            Log.This(exception);
            Environment.Exit(1);
        }

        Thread.Sleep(500);

        string response = "";

        try
        {
            response = serialPort.ReadExisting().Trim();
        }
        catch (Exception exception)
        {
            Log.This(exception);
            Environment.Exit(1);
        }

        return new Response(
                request.Id,
                int.Parse(response.Substring
                    (4, request.PinType == PinType.Digital ? 1 : 4))
            );

    }

    private Response Set(Request request)
    {

        string message = "<SD";
        if (request.PinNumber < 10) message += "0";
        message += request.PinNumber.ToString();
        message += request.Value == 0 ? "0" : "1";
        message += ">";

        try
        {
            serialPort.Write(message);
        }
        catch (Exception exception)
        {
            Log.This(exception);
            Environment.Exit(1);
        }

        Thread.Sleep(500);

        string response = "";

        try
        {
            response = serialPort.ReadExisting().Trim();
        }
        catch (Exception exception)
        {
            Log.This(exception);
            Environment.Exit(1);
        }

        return new Response(
                request.Id,
                (response.Substring(6, 2) == "OK" ?
                    Response.Result.Ok : Response.Result.Error)
            );

    }

    #endregion

}
// =============================================================================
// CLASS Schedule
// =============================================================================
class Schedule
{

    #region Properties

    public int Id { get; private set; }

    public string Port { get; private set; }

    public PinType PinType { get; private set; }

    public byte PinNumber { get; private set; }

    public int TrueValue { get; private set; }

    public int FalseValue { get; private set; }

    public ReadOnlyCollection<Condition> Conditions { get; private set; }

    #endregion

    #region Public Methods

    public static Schedule Get()
    {

        string URI = Context.ServicesRootURI +
                                "schedule/";

        XmlDocument xmlDocument;

        try
        {
            xmlDocument = REST.This(URI, "GET");
        }
        catch (Exception exception)
        {
            Log.This(URI);
            Log.This(exception);
            return null;
        }

        if (xmlDocument.GetElementsByTagName("device").Count > 0)
        {

            Schedule schedule = new Schedule();

            XmlNodeList scheduleNode = xmlDocument.GetElementsByTagName("schedule");
            schedule.Id = int.Parse(scheduleNode.Item(0).Attributes["id"].Value);

            XmlNodeList deviceNode = xmlDocument.GetElementsByTagName("device");
            foreach (XmlNode xmlNode in deviceNode.Item(0).ChildNodes)
            {
                switch (xmlNode.Name)
                {
                    case "port":
                        schedule.Port = xmlNode.InnerText;
                        break;
                    case "pinType":
                        if (xmlNode.InnerText == "A")
                            schedule.PinType = PinType.Analogue;
                        else
                            schedule.PinType = PinType.Digital;
                        break;
                    case "pinNumber":
                        schedule.PinNumber = byte.Parse(xmlNode.InnerText);
                        break;
                    case "trueValue":
                        schedule.TrueValue = int.Parse(xmlNode.InnerText);
                        break;
                    case "falseValue":
                        schedule.FalseValue = int.Parse(xmlNode.InnerText);
                        break;
                }
            }

            List<Condition> listConditions = new List<Condition>();

            XmlNodeList conditionNode = xmlDocument.GetElementsByTagName("condition");
            foreach (XmlNode xmlNode in conditionNode)
            {

                string port = "";
                PinType pinType = PinType.Analogue;
                byte pinNumber = 0;
                char sign = ' ';
                int rawValue = 0;

                foreach (XmlNode childXmlNode in xmlNode.ChildNodes)
                {
                    switch (childXmlNode.Name)
                    {
                        case "port":
                            port = childXmlNode.InnerText;
                            break;
                        case "pinType":
                            if (childXmlNode.InnerText == "A")
                                pinType = PinType.Analogue;
                            else
                                pinType = PinType.Digital;
                            break;
                        case "pinNumber":
                            pinNumber = byte.Parse(childXmlNode.InnerText);
                            break;
                        case "sign":
                            sign = childXmlNode.InnerText.ToCharArray()[0];
                            break;
                        case "rawValue":
                            rawValue = int.Parse(childXmlNode.InnerText);
                            break;
                    }

                }

                listConditions.Add(
                    new Condition(port, pinType, pinNumber, sign, rawValue)
                    );

            }

            schedule.Conditions = listConditions.AsReadOnly();

            return schedule;

        }

        return null;

    }

    public void Done(bool error)
    {

        string URI = Context.ServicesRootURI +
                                "schedule/?scheduleId=" + Id +
                                "&error=" + (error ? "Y" : "N");

        try
        {
            REST.This(URI, "POST");
        }
        catch (Exception exception)
        {
            Log.This(URI);
            Log.This(exception);
        }

    }

    #endregion

    #region Nested Class Condition

    public class Condition
    {

        public string Port { get; private set; }

        public PinType PinType { get; private set; }

        public byte PinNumber { get; private set; }

        public char Sign { get; private set; }

        public int RawValue { get; private set; }

        public Condition(
                    string port,
                    PinType pinType,
                    byte pinNumber,
                    char sign,
                    int rawValue)
        {

            Port = port;
            PinType = pinType;
            PinNumber = pinNumber;
            Sign = sign;
            RawValue = rawValue;

        }

    }

    #endregion

}
// =============================================================================
// CLASS Request
// =============================================================================
class Request
{

    #region Properties

    public int? Id { get; set; }

    public string Port { get; set; }

    public Action Action { get; set; }

    public PinType PinType { get; set; }

    public byte PinNumber { get; set; }

    public int Value { get; set; }

    #endregion

    #region Constructor

    public Request(
            int? id,
            string port,
            Action action,
            PinType pinType,
            byte pinNumber,
            int value)
    {

        Id = id;
        Port = port;
        Action = action;
        PinType = pinType;
        PinNumber = pinNumber;
        Value = value;

    }

    #endregion

    #region Public Methods

    public static Request Get()
    {

        string URI = Context.ServicesRootURI +
                        "request/";

        XmlDocument xmlDocument;

        try
        {
            xmlDocument = REST.This(URI, "GET");
        }
        catch (Exception exception)
        {
            Log.This(URI);
            Log.This(exception);
            return null;
        }

        if (xmlDocument.GetElementsByTagName("action").Item(0) != null)
        {

            Request request = new Request(
                int.Parse(
                    xmlDocument.GetElementsByTagName("requestId").Item(0).InnerText
                ),
                xmlDocument.GetElementsByTagName("port").Item(0).InnerText,
                xmlDocument.GetElementsByTagName("action").Item(0).InnerText == "G" ?
                    Action.Get : Action.Set,
                xmlDocument.GetElementsByTagName("pinType").Item(0).InnerText == "A" ?
                    PinType.Analogue : PinType.Digital,
                byte.Parse(
                    xmlDocument.GetElementsByTagName("pinNumber").Item(0).InnerText
                    ),
                int.Parse(
                    xmlDocument.GetElementsByTagName("value").Item(0).InnerText
                    )
                );

            return request;

        }

        return null;

    }

    #endregion

}
// =============================================================================
// CLASS Response
// =============================================================================
class Response
{

    #region Properties

    public int? RequestId { get; private set; }

    public string Value { get; private set; }

    #endregion

    #region Constructor

    public Response(int? requestId, int value)
    {
        RequestId = requestId;
        Value = value.ToString();
    }

    public Response(int? requestId, Response.Result result)
    {
        RequestId = requestId;
        Value = (result == Result.Error ? Value = "ER" : "OK");
    }

    #endregion

    #region Public Methods

    public void Post()
    {

        if (RequestId == null) return;

        string URI = Context.ServicesRootURI +
                        "response/?requestId=" + RequestId +
                        "&value=" + Value;

        try
        {
            REST.This(URI, "POST");
        }
        catch (Exception exception)
        {
            Log.This(URI);
            Log.This(exception);
        }

    }

    #endregion

    #region Nested Enums

    public enum Result
    {
        Error,
        Ok
    }

    #endregion

}
// =============================================================================
// CLASS REST
// =============================================================================
static class REST
{

    #region Public Methods

    public static XmlDocument This(string URI, string method)
    {

        if (URI.Contains("?"))
            URI += "&SALT=" + Context.SALT;
        else
            URI += "?SALT=" + Context.SALT;

        HttpWebRequest httpWebRequest =
                (HttpWebRequest)(WebRequest.Create(URI));
        httpWebRequest.Method = method;
        HttpWebResponse httpWebResponse =
                (HttpWebResponse)(httpWebRequest.GetResponse());
        Stream stream = httpWebResponse.GetResponseStream();
        StreamReader streamReader = new StreamReader(stream, Encoding.UTF8);

        XmlDocument xmlDocument = new XmlDocument();
        string s = streamReader.ReadToEnd();
        xmlDocument.LoadXml(s);

        return xmlDocument;

    }

    #endregion

}
// =============================================================================
// CLASS Log
// =============================================================================
static class Log
{

    #region Public Methods

    public static void This(string message)
    {
        Console.WriteLine(DateTime.Now.ToString() + " | " + message);
    }

    public static void This(Exception exception)
    {
        Log.This("[EXCEPTION] " + exception.Message);
        if (exception.InnerException != null)
            Log.This(exception.InnerException);
    }

    #endregion

}
// =============================================================================
// ENUM Action
// =============================================================================
public enum Action
{
    Get,
    Set
}
// =============================================================================
// ENUM PinType
// =============================================================================
public enum PinType
{
    Analogue,
    Digital
}
// =============================================================================